velox/functions/prestosql/aggregates/MaxSizeForStatsAggregate.cpp

```diff
@@ -0,0 +1,237 @@
+using TMaxDataSize = int64_t;
+
+class MaxSizeForStatsAggregate
+    : public SimpleNumericAggregate<TMaxDataSize, TMaxDataSize, TMaxDataSize> {
+  using BaseAggregate =
+      SimpleNumericAggregate<TMaxDataSize, TMaxDataSize, TMaxDataSize>;
+
+ public:
+  explicit MaxSizeForStatsAggregate(TypePtr resultType)
+      : BaseAggregate(resultType) {}
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(TMaxDataSize);
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    BaseAggregate::doExtractValues(groups, numGroups, result, [&](char* group) {
+      return *BaseAggregate::Aggregate::template value<TMaxDataSize>(group);
+    });
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    exec::Aggregate::setAllNulls(groups, indices);
+    for (auto i : indices) {
+      *BaseAggregate ::value<TMaxDataSize>(groups[i]) = 0;
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    // partial and final aggregations are the same
+    extractValues(groups, numGroups, result);
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    BaseAggregate::template updateGroups<true, TMaxDataSize>(
+        groups,
+        rows,
+        args[0],
+        [](TMaxDataSize& result, TMaxDataSize value) {
+          if (result < value) {
+            result = value;
+          }
+        },
+        mayPushdown);
+  }
+
+  void addSingleGroupIntermediateResults(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    BaseAggregate::updateOneGroup(
+        group,
+        rows,
+        args[0],
+        [](TMaxDataSize& result, TMaxDataSize value) {
+          result = result > value ? result : value;
+        },
+        [](TMaxDataSize& result, TMaxDataSize value, int /* unused */) {
+          result = value;
+        },
+        mayPushdown,
+        (TMaxDataSize)0);
+  }
+
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    doUpdateSingleGroup(group, rows, args[0]);
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    doUpdate(groups, rows, args[0]);
+  }
+
+ protected:
+  void upsertOneAccumulator(
+      char* const group,
+      const BaseVector& inputVector,
+      vector_size_t idx) {
+    if (inputVector.isNullAt(idx)) {
+      return;
+    }
+    // Recursively calculate size
+    size_t size_out = 0;
+    exec::MaxSizeForStatsEstimator::instance().estimateSizeOfVectorElements(
```

Comment:
Computing size one row at a time is inefficient. Let's compute for all rows at once by changing the API to take a vector and return an array of sizes (one per row). Something similar to VectorSerde::estimateSerializedSize. 

I wonder if we could re-use the logic from VectorSerde::estimateSerializedSize.

Reply:
I re-implemented this aggregate using VectorSerde::estimateSerializedSize and found that using VectorSerde::estimateSerializedSize does not produce identical reduces that we would get from Presto Java version(and thus some unit tests in this MR fail).

This is mainly due to 2 reasons:
1. For nested array types, such as array<array<bigint>>, PrestoJava(and my original implementation) do not account for storage occurred through intermediate vectors(i.e offsets and indece's).
2. For TimeStamp, Presto Java(and my original implementation) counts as 8 bytes whereas VectorSerde::estimateSerializedSize counts as 16 bytes(native size in velox).

I'd like to check, should this aggregate behave identically to Presto Java(i.e. what my original implementation does) or should it report sizes that are "more natve" to velox. Once you let me know which way we should go, I can send out another iteration.

Thanks!

Reply:
@shiyu-bytedance ANALYZE is used to collect statistics that are later used by the Cost-Based Optimizer (CBO). I don't remember off the top of my head what values CBO expects for the max-data-size. Please, do some research to figure this out.

Reply:
@mbasmanova 
I did some research(through commit history) and was able to confirm (with presto optimizer folks) that this aggregate data is not used at all by CBO.

It's a value that's collected as a "nice to have", but it's not used and there's no plan to use it.
HOWEVER, it must be supplied because otherwise ANALYZE stmt's will not run.

For this reason, I recommend to be on the conservative side:
1. sticking to my original implementation's semantics (identical with presto).
2. Incorporate your suggestion of changing the API to take a vector and return an array of sizes.

Does that sound good?

Reply:
@shiyu-bytedance Thank you for the research. Based on the results, I'd suggest to reuse VectorSerde and avoid writing and testing new code that is not going to really needed. Alternatively, we can just return zeros and have a TODO to provide meaningful values if / when Presto's optimizer starts using these.

Reply:
switched to VectorSerde.